db/db_versioned.go

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) 2024 IoTeX Foundation
+// This is an alpha (internal) release and is not suitable for production. This source code is provided 'as is' and no
+// warranties are given as to title or non-infringement, merchantability or fitness for purpose and, to the extent
+// permitted by law, all liability for your use of the code is disclaimed. This source code is governed by Apache
+// License 2.0 that can be found in the LICENSE file.
+
+package db
+
+import (
+	"context"
+)
+
+type (
+	// KvVersioned is a versioned key-value store, where each key can (but doesn't
+	// have to) have multiple versions of value (corresponding to different heights
+	// in a blockchain)
+	//
+	// A namespace/bucket is considered versioned by default, user needs to use the
+	// NonversionedNamespaceOption() to specify non-versioned namespaces at the time
+	// of creating the versioned key-value store
+	//
+	// Versioning is achieved by using (key + 8-byte version) as the actual storage
+	// key in the underlying DB. For buckets containing versioned keys, a metadata
+	// is stored at the special key = []byte{0}. The metadata specifies the bucket's
+	// name and the key length
+	//
+	// For each versioned key, the special location = key + []byte{0} stores the
+	// key's metadata, which includes the following info:
+	// 1. the version when the key is first created
+	// 2. the version when the key is lastly written
+	// 3. the version when the key is deleted
+	// 4. hash of the key's last written value
+	// If the location does not store a value, the key has never been written.
+	//
+	// Here's an example of a versioned DB which has 3 buckets:
+	// 1. "mta" -- regular bucket storing metadata, key is not versioned
+	// 2. "act" -- versioned namespace, key length = 20
+	// 3. "stk" -- versioned namespace, key length = 8
+	KvVersioned interface {
+		KVStore
+
+		// Version returns the key's most recent version
+		Version(string, []byte) (uint64, error)
+
+		// SetVersion sets the version before calling Put()
+		SetVersion(uint64) KvVersioned
+
+		// CreateVersionedNamespace creates a namespace to store versioned keys
+		CreateVersionedNamespace(string, uint32) error
+	}
+
+	// BoltDBVersioned is KvVersioned implementation based on bolt DB
+	BoltDBVersioned struct {
+		*BoltDB
+		version      uint64          // version for Get/Put()
+		versioned    map[string]int  // buckets for versioned keys
```

Comment:
If there are many keys, will these two maps be very large?

Reply:
i think the map key should be namespace instead of key, is it right?

Reply:
yes, the map key is the namespace